.github/workflows/merge.yaml

```diff
@@ -0,0 +1,125 @@
+name: "GitHub CI merge checks"
+
+on:
+  push:
+    branches: [ "auto" ]
+
+env:
+  PULL_REQUEST_NUMBER: ${{ github.event.pull_request.number }}
+  CHECKOUT_FETCH_DEPTH: 1001
+
+jobs:
+  linux-matrix-tests:
+
+    strategy:
+      matrix:
+        os:
+          - centos-stream-9
+          - debian-stable
+          - debian-testing
+          - debian-unstable
+          - fedora-39
+          - fedora-40
+          - fedora-rawhide
+          - gentoo
+          - opensuse-leap
+          - opensuse-tumbleweed
+          - ubuntu-focal
+          - ubuntu-jammy
+          - ubuntu-noble # EOL 2036-04
+          - ubuntu-oracular # EOL 2025-07
+        compiler:
+          - { CC: gcc, CXX: g++ }
+          - { CC: clang, CXX: clang++ }
+        layer:
+          - { name: layer-00-default, nick: default }
+          - { name: layer-01-minimal, nick: minimal }
+          - { name: layer-02-maximus, nick: maximus }
+        # exclude:
+        #  if any combinations needs to be excluded, do it here
+        #   - { os: macos-14, compiler: { CC: gcc, CXX: g++ } }
+
+    runs-on: ubuntu-latest
+    container:
+      image: squidcache/buildfarm-${{ matrix.os }}:stable
+      options: --user 1001 # uid used by worfklow runner
+
+    name:  linux-matrix-test(${{ matrix.os }},${{ matrix.compiler.CC }},${{ matrix.layer.nick }})
+    env:
+      CC: ${{ matrix.compiler.CC }}
+      CXX: ${{ matrix.compiler.CXX }}
+
+    steps:
+      - name: Checkout Sources
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: ${{ env.CHECKOUT_FETCH_DEPTH }}
+
+      - name: Run build test
+        id: test
+        run: |
+          ./test-builds.sh --verbose --use-config-cache ${{ matrix.layer.name }}
+
+      - name: Publish build logs
+        if: success() || failure()
+        uses: actions/upload-artifact@v4
+        with:
+          name: build-logs-${{ matrix.os }}-${{ matrix.compiler.CC }}-${{ matrix.layer.nick }}
+          path: btlayer-*.log
+
+  macos:
+    runs-on: macos-14
+    name: macos-test
+
```

Comment:
Why? This all is unnecessary complication, it buys us nothing at all.
We do not automate artifacts collection. If we ever will want a machine
to do touch these artifacts, that will be the moment to consider
whether to standardise.

Also, this workflow is working just fine, I wouldn't characterise any of the
suggestions in this comment as "problems".

Reply:
Ah right.
The answer is pretty simple then: simplicity.
The setup phase on homebrew is quite different from Linux's.
As this PR adds significant complexity by having a large number of matrix combinations, I decided it was best to isolate MacOS complexity on its own.
We could possibly parallelise macos tests but I see little benefit in doing that for now.

In another thread you mentioned that you assume there will be more platforms tested, and that is indeed correct: the FreeBSD workflow is working and will can discuss it as soon as this PR lands

Reply:
> I would even use os/compiler/layer matrix (with two components fixed) so that we can use the same artifact name scheme and naming code instead of unnecessary special-casing MacOS in that way.

Just to avoid misunderstanding: This change request does _not_ ask to merge MacOS and non-MacOS tests into _one_ common matrix (and `if` statements to differentiate OSes). MacOS tests should use their own matrix.

To recap: If docker tests iterate over build layers (among other matrix dimensions), then MacOS tests should also iterate over build layers (at least). Otherwise (i.e. if docker tests are changed to stop iterating over build layers), MacOS tests should not iterate over build layers either.